### ISSUE TITLE
[bug] Memory leak after upgrading mongoose from version 4 to 5

### ISSUE THREAD

**Issue by aenavcam** [Automattic/mongoose#7223]
We recently updated mongoose to version 5 in three of our services and we're experiencing what it seems to be a memory leak. It happens very slowly and it's independent from the activity of the service running it. We've tried looking at heap snapshots but nothing seems obviously suspicious. Any help is appreciated.
Node 8.12.0
MongoDB 3.6.4
Mongoose 5.3.6

![image-2018-10-22-17-34-23-247](https://user-images.githubusercontent.com/10712656/48058441-b7474980-e1b6-11e8-9171-838dbab8ab99.png)

this is the code that initialises the connection, it runs only once and make the mongoose instance available as a singleton:

```
		const mongoose = require( 'mongoose' );
		const Scanner = require( './scanner' );
		const path = require( 'path' );
		const Promise = require( 'bluebird' );
		mongoose.plugin( require( 'mongoose-paginate' ) );

		service.registerSingleton( 'mongoose', mongoose );

		mongoose.Promise = Promise;
		// scan directory and import models
		mongoose.models = modelPath ? Scanner.import( Scanner.scan( path.join( modelPath, 'mongo', '**', '*.js' ) ) ) : {};

		const serviceMongooseOptions = mongoConfig.options || {};
		const mongooseOptions = defaultsDeep(
				serviceMongooseOptions,
				{ connectTimeoutMS: 10000, socketTimeoutMS: 10000 }
			);

		const connection = mongoConfig.url ?
			mongoose.connect(
				mongoConfig.url,
				mongooseOptions
			) :
			Promise.reject( new Error( `Mongoose#connect requires a url.` ) );

		return connection;
```

**Comment by NewEraCracker** [comment]
Have you tried with an earlier version such as 5.0.18 or 5.1.8 to see if the behaviour also happens there?

**Comment by aenavcam** [comment]
I've just deployed with 5.0.8, I'll report back

**Comment by vkarpov15** [comment]
Can you provide any more info on what your code looks like and what your services do?

**Comment by ericlacher** [comment]
Please see this chart showing how the different versions affected our memory consumption and cpu usage. We have 3 services using mongoose, and the code we use to initialise it is what @aenavcam already showed you in the initial post. Please tell me what else could help you there.
The services are just idling, as long as we do not actively trigger them (this is our dev env, so we have control about any calls made).
![mongoose_cpu](https://user-images.githubusercontent.com/11710592/48415388-8edac480-e74d-11e8-839a-2142486661be.png)
It is quite complicated to replace the mongoose version, as we have a framework dep, which has a library dep, which has the mongoose dep, so quite a lot of packaging and branching needed, that's why we not yet tested 5.1.18. But if you absolutely need to know, I will do that.
Please also note that nothing was changed, except mongoose version.

**Comment by vkarpov15** [comment]
@ericlacher can you clarify what framework you're using? Also, what do you mean by "services are just idling, as long as we do not actively trigger them", does that mean the services just maintain a connection to MongoDB and don't do anything else?

**Comment by ericlacher** [comment]
@vkarpov15 framework is hapi.js with homegrown surroundings. Yes, your assumption is correct, the services just sit there without anyone issuing a query. Only exception is grafana fetching metrics with prometheus plugin, but I disabled this for a test round and the result was the same.

**Comment by aenavcam** [comment]
I've also noticed that the latest version of mongoose seems to call the isMaster command every 5 seconds and after a while the bson/decrypt seems to leave a couple of arrays hanging.

**Comment by vkarpov15** [comment]
Did some more digging today, still no progress to report. I will continue working on this, but any help or insights are most welcome.

**Comment by vkarpov15** [comment]
I looked into this a little more and I'm more confused than before, because even the below script shows monotonic memory growth. So the memory growth from the script I ran with Mongoose doesn't seem useful.

```javascript
const start = Date.now();

print();

function print() {
  console.log(`+${Date.now() - start}`, process.memoryUsage().heapUsed);
  setTimeout(print, 5000);
}
```

```
$ node gh-7223-no-op.js 
+0 4406904
+5008 4727448
+10012 4736544
+15017 4738552
+20021 4740600
+25027 4742408
+30032 4744216
+35038 4746264
+40042 4747928
+45048 4749592
+50054 4751496
+55059 4753160
+60061 4754824
+65065 4756728
+70069 4758392
+75073 4760056
+80077 4761960
+85081 4763624
+90087 4765288
+95089 4767192
+100095 4768856
+105101 4770520
+110106 4772424
+115111 4774088
+120117 4775752
+125123 4777656
+130126 4779320
+135130 4780984
+140135 4782888
+145142 4784552
+150147 4786216
+155151 4788120
+160157 4789784
+165163 4791448
+170167 4793352
+175172 4795016
+180177 4796680
+185183 4798584
+190188 4800248
+195195 4801912
+200196 4803816
+205201 4805480
+210206 4807144
+215208 4809048
+220214 4810712
+225215 4812376
```

I'm gonna need some more info as to:

1) How you're tracking memory usage
2) What the frameworks/libs in "a framework dep, which has a library dep, which has the mongoose dep" are

**Comment by WilliamHuang1995** [comment]
any update on this?

**Comment by vkarpov15** [comment]
@WilliamHuang1995 no, I haven't been able to confirm that there is any memory leak. Please open up a new issue and follow the issue template, bump comments are not helpful.
